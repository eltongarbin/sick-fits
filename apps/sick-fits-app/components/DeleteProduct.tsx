import { useMutation } from '@apollo/client';
import gql from 'graphql-tag';
import { ReactNode } from 'react';

const DELETE_PRODUCT_MUTATION = gql`
  mutation DELETE_PRODUCT_MUTATION($id: ID!) {
    deleteProduct(where: { id: $id }) {
      id
      name
    }
  }
`;

type DeleteProductProps = {
  id: string;
  children: ReactNode;
};

export const DeleteProduct = ({ id, children }: DeleteProductProps) => {
  const [deleteProduct, { loading }] = useMutation(DELETE_PRODUCT_MUTATION, {
    variables: { id },
    update(cache, { data: { deleteProduct } }) {
      cache.evict({ id: cache.identify(deleteProduct) });
    },
  });

  return (
    <button
      type="button"
      disabled={loading}
      onClick={() => {
        if (confirm('Are you sure you want to delete this item?')) {
          deleteProduct().catch((err) => alert(err.message));
        }
      }}
    >
      {children}
    </button>
  );
};
