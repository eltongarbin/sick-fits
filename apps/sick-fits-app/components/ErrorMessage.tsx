import { ApolloError, ServerError } from '@apollo/client';
import styled from 'styled-components';

const ErrorStyles = styled.div`
  padding: 2rem;
  background: white;
  margin: 2rem 0;
  border: 1px solid rgba(0, 0, 0, 0.05);
  border-left: 5px solid red;

  p {
    margin: 0;
    font-weight: 100;
  }

  strong {
    margin-right: 1rem;
  }
`;

function isServerError(
  networkError: ApolloError['networkError']
): networkError is ServerError {
  return (networkError as ServerError)?.result !== undefined;
}

type DisplayErrorProps = {
  error?: ApolloError;
};

export const ErrorMessage = ({ error }: DisplayErrorProps) => {
  if (!error || !error.message) return null;

  if (
    isServerError(error.networkError) &&
    error.networkError.result.errors.length
  ) {
    return error.networkError.result.errors.map((error: Error, i: number) => (
      <ErrorStyles key={i}>
        <p data-testid="graphql-error">
          <strong>Shoot!</strong>
          {error.message.replace('GraphQL error: ', '')}
        </p>
      </ErrorStyles>
    ));
  }

  return (
    <ErrorStyles>
      <p data-testid="graphql-error">
        <strong>Shoot!</strong>
        {error.message.replace('GraphQL error: ', '')}
      </p>
    </ErrorStyles>
  );
};
