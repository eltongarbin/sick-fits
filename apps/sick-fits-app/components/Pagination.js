import gql from 'graphql-tag';
import Head from 'next/head';
import Link from 'next/link';
import { useQuery } from '@apollo/client';

import PaginationStyles from './styles/PaginationStyles';
import DisplayError from './ErrorMessage';
import { perPage } from '../config';

export const PAGINATION_QUERY = gql`
  query PAGINATION_QUERY {
    productsCount
  }
`;

export default function Pagination({ page }) {
  const { data, loading, error } = useQuery(PAGINATION_QUERY);

  if (loading) return 'Loading...';
  if (error) return <DisplayError error={error} />;

  const count = data.productsCount;
  const pageCount = Math.ceil(count / perPage);

  return (
    <PaginationStyles data-testid="pagination">
      <Head>
        <title>
          Sick Fits - Page {page} of {pageCount}
        </title>
      </Head>
      <Link href={`/products/${page - 1}`} aria-disabled={page <= 1}>
        ← Prev
      </Link>
      <p>
        Page {page} of {pageCount}
      </p>
      <p>{count} Items Total</p>
      <Link href={`/products/${page + 1}`} aria-disabled={page >= pageCount}>
        Next →
      </Link>
    </PaginationStyles>
  );
}
