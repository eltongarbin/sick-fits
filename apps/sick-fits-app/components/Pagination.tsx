import gql from 'graphql-tag';
import Head from 'next/head';
import Link from 'next/link';
import { useQuery } from '@apollo/client';

import { PaginationStyles } from './Pagination.styles';
import { ErrorMessage } from './ErrorMessage';
import { perPage } from '../config';

export const PAGINATION_QUERY = gql`
  query PAGINATION_QUERY {
    productsCount
  }
`;

type PaginationProps = {
  page: number;
};

export const Pagination = ({ page }: PaginationProps) => {
  const { data, loading, error } = useQuery(PAGINATION_QUERY);

  if (loading) return <>Loading...</>;
  if (error) return <ErrorMessage error={error} />;

  const count = data.productsCount;
  const pageCount = Math.ceil(count / perPage);

  return (
    <PaginationStyles data-testid="pagination">
      <Head>
        <title>
          Sick Fits - Page {page} of {pageCount}
        </title>
      </Head>
      <Link href={`/products/${page - 1}`} aria-disabled={page <= 1}>
        ← Prev
      </Link>
      <p>
        Page {page} of {pageCount}
      </p>
      <p>{count} Items Total</p>
      <Link href={`/products/${page + 1}`} aria-disabled={page >= pageCount}>
        Next →
      </Link>
    </PaginationStyles>
  );
};
