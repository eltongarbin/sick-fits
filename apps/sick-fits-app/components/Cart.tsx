import { CloseButton } from './styles/CloseButton';
import { CartStyles } from './styles/CartStyles';
import { Supreme } from './styles/Supreme';
import { useUser } from './User';
import { formatMoney } from '../lib/formatMoney';
import { calcTotalPrice } from '../lib/calcTotalPrice';
import { useCart } from '../lib/cartState';
import { Checkout } from './Checkout';
import { CartItem } from './CartItem';

export const Cart = () => {
  const me = useUser();
  const { cartOpen, closeCart } = useCart();

  if (!me) return null;

  return (
    <CartStyles open={cartOpen}>
      <header>
        <Supreme>{`${me.name}'s Cart`}</Supreme>
        <CloseButton onClick={closeCart}>&times;</CloseButton>
      </header>
      <ul>
        {me.cart.map((cartItem: any) => (
          <CartItem key={cartItem.id} cartItem={cartItem} />
        ))}
      </ul>
      <footer>
        <p>{formatMoney(calcTotalPrice(me.cart))}</p>
        <Checkout />
      </footer>
    </CartStyles>
  );
};
