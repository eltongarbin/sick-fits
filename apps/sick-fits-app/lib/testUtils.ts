import casual from 'casual';
import { PAGINATION_QUERY } from '../components/Pagination';

// seed it so we get consistent results
casual.seed(777);

const fakeItem = () => ({
  // __typename: 'Item',
  id: 'abc123',
  price: 5000,
  user: null,
  photo: {
    image: {
      publicUrlTransformed: 'dog.jpg',
    },
    altText: 'dogs are best',
  },
  name: 'dogs are best',
  description: 'dogs',
});

const fakeUser = (overrides?: Record<string, unknown>) => ({
  __typename: 'User',
  id: '4234',
  name: casual.name,
  email: casual.email,
  permissions: ['ADMIN'],
  orders: [],
  cart: [],
  ...overrides,
});

const fakeOrderItem = () => ({
  __typename: 'OrderItem',
  id: casual.uuid,
  image: {
    image: `${casual.word}.jpg`,
  },
  name: casual.words(),
  price: 4234,
  quantity: 1,
  description: casual.words(),
});

const fakeOrder = () => ({
  __typename: 'Order',
  id: 'ord123',
  charge: 'ch_123',
  total: 40000,
  items: [fakeOrderItem(), fakeOrderItem()],
  createdAt: '2022-12-11T20:16:13.797Z',
  user: fakeUser(),
});

const fakeCartItem = (overrides?: Record<string, unknown>) => ({
  __typename: 'CartItem',
  id: 'omg123',
  quantity: 3,
  product: fakeItem(),
  user: fakeUser(),
  ...overrides,
});

// Fake LocalStorage
class LocalStorageMock {
  store: Record<string, string> = {};

  clear() {
    this.store = {};
  }

  getItem(key: string) {
    return this.store[key] || null;
  }

  setItem(key: string, value: string) {
    this.store[key] = value.toString();
  }

  removeItem(key: string) {
    delete this.store[key];
  }
}

const makePaginationMocksFor = (length: number) => [
  {
    request: { query: PAGINATION_QUERY },
    result: {
      data: {
        productsCount: length,
      },
    },
  },
];

export {
  makePaginationMocksFor,
  LocalStorageMock,
  fakeItem,
  fakeUser,
  fakeCartItem,
  fakeOrder,
  fakeOrderItem,
};
