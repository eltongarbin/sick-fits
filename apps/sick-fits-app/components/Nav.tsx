import Link from 'next/link';

import { useCart } from '../lib/cartState';
import { CartCount } from './CartCount';
import { SignOut } from './SignOut';
import { NavStyles } from './Nav.styles';
import { useUser } from './User';

export const Nav = () => {
  const user = useUser();
  const { openCart } = useCart();

  return (
    <NavStyles>
      <Link href="/products">Products</Link>
      {user && (
        <>
          <Link href="/sell">Sell</Link>
          <Link href="/orders">Orders</Link>
          <SignOut />
          <button type="button" onClick={openCart}>
            My Cart
            <CartCount
              count={user.cart.reduce(
                (tally: number, cartItem: any) =>
                  tally + (cartItem.product ? cartItem.quantity : 0),
                0
              )}
            />
          </button>
        </>
      )}
      {!user && <Link href="/signin">Sign In</Link>}
    </NavStyles>
  );
};
