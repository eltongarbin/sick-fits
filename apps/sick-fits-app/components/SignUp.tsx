import gql from 'graphql-tag';
import { useMutation } from '@apollo/client';

import { Form } from './styles/Form';
import { useForm } from '../lib/useForm';
import { ErrorMessage } from './ErrorMessage';
import { SyntheticEvent } from 'react';

export const SIGNUP_MUTATION = gql`
  mutation SIGNUP_MUTATION(
    $name: String!
    $email: String!
    $password: String!
  ) {
    createUser(data: { name: $name, email: $email, password: $password }) {
      id
      email
      name
    }
  }
`;

export const SignUp = () => {
  const { inputs, handleChange, resetForm }: any = useForm({
    name: '',
    email: '',
    password: '',
  });
  const [signup, { data, error }] = useMutation(SIGNUP_MUTATION, {
    variables: inputs,
  });

  const handleSubmit = async (e: SyntheticEvent) => {
    e.preventDefault();
    await signup().catch(console.error);
    resetForm();
  };

  return (
    <Form method="POST" onSubmit={handleSubmit}>
      <h2>Sign Up For an Account</h2>
      <ErrorMessage error={error} />
      <fieldset>
        {data?.createUser && (
          <p>
            Signed up with {data.createUser.email} - Please Go Head and Sign in!
          </p>
        )}
        <label htmlFor="name">
          Your Name
          <input
            id="name"
            type="text"
            name="name"
            placeholder="Your Name"
            autoComplete="name"
            value={inputs.name}
            onChange={handleChange}
          />
        </label>
        <label htmlFor="email">
          Email
          <input
            id="email"
            type="email"
            name="email"
            placeholder="Your Email Address"
            autoComplete="email"
            value={inputs.email}
            onChange={handleChange}
          />
        </label>
        <label htmlFor="password">
          Password
          <input
            id="password"
            type="password"
            name="password"
            placeholder="Password"
            autoComplete="password"
            value={inputs.password}
            onChange={handleChange}
          />
        </label>
        <button type="submit">Sign Up!</button>
      </fieldset>
    </Form>
  );
};
