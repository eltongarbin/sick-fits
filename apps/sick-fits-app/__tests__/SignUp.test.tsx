import { render, screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/client/testing';
import userEvent from '@testing-library/user-event';

import { SignUp, SIGNUP_MUTATION } from '../components/SignUp';
import { CURRENT_USER_QUERY } from '../components/User';
import { fakeUser } from '../lib/testUtils';

const me = fakeUser();
const password = 'test';
const mocks = [
  {
    request: {
      query: SIGNUP_MUTATION,
      variables: { name: me.name, email: me.email, password },
    },
    result: {
      data: {
        createUser: {
          __typename: 'User',
          id: 'abc123',
          email: me.email,
          name: me.name,
        },
      },
    },
  },
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { authenticatedItem: me } },
  },
];

describe('<SignUp />', () => {
  it('renders and matches snapshot', () => {
    const { container } = render(
      <MockedProvider>
        <SignUp />
      </MockedProvider>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('calls the mutation properly', async () => {
    render(
      <MockedProvider mocks={mocks}>
        <SignUp />
      </MockedProvider>
    );

    const user = userEvent.setup();
    await user.type(screen.getByLabelText('Your Name'), me.name);
    await user.type(screen.getByLabelText('Email'), me.email);
    await user.type(screen.getByLabelText('Password'), password);
    await user.click(screen.getByRole('button', { name: 'Sign Up!' }));

    await screen.findByText(
      `Signed up with ${me.email} - Please Go Head and Sign in!`
    );
  });
});
