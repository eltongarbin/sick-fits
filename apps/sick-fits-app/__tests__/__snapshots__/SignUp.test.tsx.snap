// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignUp /> renders and matches snapshot 1`] = `
<form
  class="sc-bcXHqe cYSsYe"
  method="POST"
>
  <h2>
    Sign Up For an Account
  </h2>
  <fieldset>
    <label
      for="name"
    >
      Your Name
      <input
        autocomplete="name"
        id="name"
        name="name"
        placeholder="Your Name"
        type="text"
        value=""
      />
    </label>
    <label
      for="email"
    >
      Email
      <input
        autocomplete="email"
        id="email"
        name="email"
        placeholder="Your Email Address"
        type="email"
        value=""
      />
    </label>
    <label
      for="password"
    >
      Password
      <input
        autocomplete="password"
        id="password"
        name="password"
        placeholder="Password"
        type="password"
        value=""
      />
    </label>
    <button
      type="submit"
    >
      Sign Up!
    </button>
  </fieldset>
</form>
`;
